### PR TITLE
Support for differentiation of parameters in DSL

### PR DESCRIPTION
**Pull request by richfitz**
This PR builds on #294 to add control over differentiation of parameters to odin. The syntax for users just replaces

```
scale <- user()
```

with

```
scale <- user(differentiate = TRUE)
```

### CODE REVIEW

File: R/ir_parse.R [hunk 1]
```diff
@@ -1063,12 +1092,23 @@ ir_parse_expr_rhs_user <- function(rhs, line, source) {
...
+
+  ## NOTE: the 'dim' part here is not actually known yet!
   user <- list(default = m$default,
                dim = FALSE,
-               integer = m$integer %||% FALSE,
+               integer = integer,
                min = m$min,
-               max = m$max)
+               max = m$max,
+               differentiate = differentiate)
```
Comment (r-ash) [inline R/ir_parse.R:1110]: Where will this be used?

**Review by r-ash — COMMENTED** [review]
Code changes themselves look fine, but I'm a bit confused. Where will this be used? What does `differentiate` mean here? You said in the epic this is for "indicating which parameters we want the likelihood differentiated with respect to". The likelihood is the compare function? Or the compare function is used to build the likelihood after running the adjoint model? How does setting this `differentiate` flag relate to that? I thought `user` was just to let users specify the parameter value used in the model. Is it possible to work this out automatically by looking at the parameters that the compare function uses?

File: tests/testthat/test-run-general.R [hunk 2]
```diff
@@ -813,7 +813,7 @@ test_that_odin("overlapping graph", {
     list(y * p, p + p2)
   }
   cmp <- deSolve::ode(1, tt, f, NULL)
-  tol <- variable_tolerance(mod, js = 1e-6)
+  tol <- variable_tolerance(mod, js = 3e-6)
```
Comment (weshinsley) [inline tests/testthat/test-run-general.R:816]: Is this intentional?

**Review by weshinsley — COMMENTED** [review]
I think all good but a couple of comments I'll ask about when you're back!

File: tests/testthat/test-run-general.R (hunk 2 above)
Comment (richfitz) [reply tests/testthat/test-run-general.R:816]: unrelated, failing on master atm. I think some change in libv8, as I don't see the failure locally

**Comment by richfitz** [comment]
> Code changes themselves look fine, but I'm a bit confused. Where will this be used? What does differentiate mean here? You said in the epic this is for "indicating which parameters we want the likelihood differentiated with respect to". The likelihood is the compare function? Or the compare function is used to build the likelihood after running the adjoint model? How does setting this differentiate flag relate to that? I thought user was just to let users specify the parameter value used in the model. Is it possible to work this out automatically by looking at the parameters that the compare function uses?

Only `user()` parameters can be used as the thing we differentiate with respect to, but not all can. For example, in the motivating SIR model we can't differentiate against `freq` as it is an integer, and we probably would not want to differentiate with respect to `dt` if we provided that directly. Other things we might not differentiate against would be user "parameters" that represent data (mixing matrices, vaccination schedules etc).

The compare function goes into the likelihood (the likelihood is the sum over all the times that the compare gets called). The parameters that we differentiate with respect to get added into the adjoint model so that it has some extra variables, each corresponding to the partial derivative of the likelihood with respect to a parameter.

All parameters will ultimately end up somewhere in the graph, and it's not just the bit in the compare function that matters. Again, in the motivating SIR model we differentiate with respect to parameters that only appear in the forward time model and *not* in the compare function

File: R/ir_parse.R (hunk 1 above)
Comment (richfitz) [reply R/ir_parse.R:1110]: in odin.dust, because this property makes it out in the IR

**Review by r-ash — APPROVED** [review]

**Review by weshinsley — APPROVED** [review]
Looks good!

**Pull request merged**